lib/ngx/re.lua

```diff
@@ -0,0 +1,77 @@
+-- Copyright (C) Thibault Charbonnier (thibaultcha)
+-- I hereby assign copyright in this code to the lua-resty-core project,
+-- to be licensed under the same terms as the rest of the code.
+
+
+local base = require "resty.core.base"
+local regex = require "resty.core.regex"
+
+
+local re_match_helper = regex.re_match_helper
+local tostring = tostring
+local new_tab = base.new_tab
+local sub = string.sub
+
+
+local _M = { version = base.version }
+
+
+function _M.split(subj, regex, opts, ctx, max, nth, res)
+    -- we need to cast this to strings to avoid exceptions when they are
+    -- something else.
+    -- needed because of further calls to string.sub in this function.
+    subj = tostring(subj)
+
+    if not ctx then
+        ctx = new_tab(0, 1)
+    end
+
+    if not ctx.pos then
+        ctx.pos = 1
+    end
+
+    if not res then
+        res = new_tab(0, 0)
+    elseif type(res) ~= "table" then
+        return error("res is not a table", 2)
+    end
+
+    local idx = ctx.pos
+    local n = 0
+
+    while true do
+        local from, to, err, caps = re_match_helper(subj, regex, opts, ctx, nil, true)
+        if err then
+            return nil, err
+        elseif not from then
```

Comment:
The previous branch always returns, so we should not use `elseif` or `if` here for the 2nd branch.